server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -292,6 +292,74 @@ public ShardIterator activeInitializingShardsRankedIt(
         return new PlainShardIterator(shardId, ordered);
     }
 
+    /**
+     * Returns an iterator over active and initializing shards, shards are ordered by weighted round-robin scheduling
+     * policy.
+     * @param wrrWeight Weighted round-robin weight entity
+     * @param nodes discovered nodes in the cluster
+     * @return an iterator over active and initializing shards, ordered by weighted round-robin
+     * scheduling policy. Making sure that initializing shards are the last to iterate through.
+     */
+    public ShardIterator activeInitializingShardsWRR(WRRWeights wrrWeight, DiscoveryNodes nodes, WRRShardsCache cache) {
+        final int seed = shuffler.nextSeed();
+        List<ShardRouting> ordered = new ArrayList<>(activeShards.size() + allInitializingShards.size());
+        List<ShardRouting> orderedActiveShards;
+        if (cache.get(new WRRShardsCache.Key(shardId)) != null) {
+            orderedActiveShards = cache.get(new WRRShardsCache.Key(shardId));
+        } else {
+            orderedActiveShards = getShardsWRR(activeShards, wrrWeight, nodes);
+            cache.put(new WRRShardsCache.Key(shardId), orderedActiveShards);
+        }
```

Comment:
All mutating cache operations needs to be thread safe. Can we please ensure we handle concurrency

Reply:
This is handled in OpenSearch [Cache](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/cache/Cache.java) implementation which we are using for caching.

Reply:
No not fully. For instance between the two successive cache gets its possible that the cache would have been invalidated which would lead to concurrency bugs.

Reply:
addressed

Reply:
The concern is that if the cache is empty multiple threads would try to compute the weights and then update it. This work might be wasteful and can lead to races. Can we synchronise the computation